### PR TITLE
Assembler module

### PR DESCRIPTION
**Pull request by cristeahub**
This prq adds the assembler module. This module has all the functions needed to be able to assemble the IR to hex code. What is not included here is the part needed to connect this to the rest of the main module.

This is an iteration on #20 .

### CODE REVIEW

File: src/Opcodes.hs [hunk 1]
```diff
@@ -14,7 +15,54 @@ data Opcode = And
...
+            | Sw
+            | ThreadFinished
     deriving (Show, Eq)
 
 mnemonic :: Opcode -> String
 mnemonic = map toLower . show
+
+hexOpcode :: Opcode -> Int
+hexOpcode opcode =
+  case opcode of
```
Comment (aleksanb) [inline src/Opcodes.hs]: can change this to
code `shiftL` 26
where code = case opcode of, so we only need one shift!

File: src/Assembler.hs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+                BitwiseOr -> Or
+                BitwiseXor -> Xor
+                Plus -> Add
+                Minus -> Sub
+                Multiply -> Mul
+                LessThan -> Slt
+                EqualTo -> Seq
+
+assembleInstruction (ThreeIR op rd (I rs) (R rt) masked)
+  | op == Plus = assembleInstruction (ThreeIR op rd (R rt) (I rs) masked)
```
Comment (aleksanb) [inline src/Assembler.hs]: Should the flip only be done for the Plus operation?

File: src/Assembler.hs [hunk 3]
```diff
@@ -0,0 +1,43 @@
+module Assember(assemble) where
```
Comment (aleksanb) [inline src/Assembler.hs]: Should be assemb_l_er

File: src/Assembler.hs (hunk 2 above)
Comment (cristeahub) [inline src/Assembler.hs]: This is taken from the Serializer file https://github.com/aleksanb/hdc/blob/master/src/Serializer.hs#L31 .

**Comment by cristeahub** [comment]
ping!

**Comment by aleksanb** [comment]
Looks good. Only thing missing is that address and immediate values need to be taken modulo 2**16!

**Comment by cristeahub** [comment]
ping!

**Pull request merged**